### PR TITLE
Refactor internals and improve RSpec support

### LINKED ISSUE #15: RSpec integration not working as expected
I've currently setup `packwerk` and `stimpack` in my application and my expectation around RSpec was that when running `rspec` or `bundle exec rspec` then all tests contained in the standard `spec/` folder and all tests in `packs/**/spec/` folders would be executed.

However it seems like at the moment only tests in the global folder are being picked up and executed, the packs tests are being ignored. I'm using the new `Stimpack.config.root` setting but I don't think that has anything to do with this.

While debugging, I've managed to print out the value of `RSpec.configuration.pattern` after the Stimpack integration was loaded, and I got the following:

```
"**{,/*/**}/*_spec.rb,components/comp1/spec/**/*_spec.rb,components/comp2/spec/**/*_spec.rb,components/comp3/spec/**/*_spec.rb"
```

so I can see my components there, but the tests are still not included in the run.
Is my expectation incorrect, or is this a bug with Stimpack?

For what is worth, I was previously able to make this work in my app before adding Stimpack, and the trick had to do with setting the `--default-path` option for RSpec.
This is because, as far as I understand, RSpec assumes your path starts from `spec/` and so all patterns are relative to that path, hence `components/...` will translate into `spec/components...` which is not what we want.

### PR DESCRIPTION
**Pull request by ngan**
This PR refactors Stimpack and breaks the initialization step into 2 process: non-Rails and Rails. The non-rails part simply provides packs and pack traversal. When Rails is involved, we inject the packs into Rails (along with Engine support). This helps us extract the RSpec helper so that it works without Rails booting.

Fixes #15

For better RSpec support, you now have to `--require stimpack/rspec`. You can add this to `.rspec` or add it to your command line: `bundle exec rspec --require stimpack/rspec ...`

This allows Stimpack to hook into RSpec before files are resolved. Because of this, we're now able to do some fancy stuff:

(This assumes you've added `--require stimpack/rspec` to `.rspec`)
```
bundle exec rspec #=> Runs _all_ specs in your entire application (packs and all)
bundle exec rspec spec/some/specific_spec.rb #=> Runs just that one test
bundle exec rspec packs/foobar #=> Runs all specs under the foobar pack _and_ all the specs of its nested packs
bundle exec rspec packs/foobar pack/binbaz #=> same as above but also adds the binbaz pack
bundle exec rspec packs/foobar/spec #=> Runs all spec files inside the "packs/foobar/spec" directory
bundle exec rspec packs/foobar/nested_pack #=> Runs all specs under the "packs/foobar/nested_pack" pack
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -92,6 +92,14 @@ metadata:
   engine: true
 ```
 
+### RSpec Integration
+It's highly recommended that you add `--require stimpack/rspec` to your `.rspec` file to keep things simply.
```
Comment (AlexEvanczuk) [inline README.md:96]:
```suggestion
It's highly recommended that you add `--require stimpack/rspec` to your `.rspec` file to keep things simple.
```

File: spec/stimpack_spec.rb [hunk 2]
```diff
@@ -1,6 +1,7 @@
 require "pathname"
 
 rails_dir = Pathname.new(File.expand_path("fixtures/rails-7.0", __dir__))
+Dir.chdir(rails_dir)
```
Comment (AlexEvanczuk) [inline spec/stimpack_spec.rb:4]: Why is this necessary?

Also -- I wonder if we should have a `use_fixture` helper which loads the right rails environment. 

Separately -- I realize we might need to have separate rspec processes that load different rails environments, unless there is a way to "unload" a rails environment.

**Comment by AlexEvanczuk** [comment]
> For better RSpec support, you now have to --require stimpack/rspec.

"You now have to" => "You now can" ?
It's unclear to me if this is *required* for rspec support or if not, under what conditions it's recommended, and what it adds.

```
bundle exec rspec #=> Runs _all_ specs in your entire application (packs and all)
bundle exec rspec spec/some/specific_spec.rb #=> Runs just that one test
bundle exec rspec packs/foobar #=> Runs all specs under the foobar pack _and_ all the specs of its nested packs
bundle exec rspec packs/foobar pack/binbaz #=> same as above but also adds the binbaz pack
bundle exec rspec packs/foobar/spec #=> Runs all spec files inside the "packs/foobar/spec" directory
bundle exec rspec packs/foobar/nested_pack #=> Runs all specs under the "packs/foobar/nested_pack" pack
```

I think this would actually be great to add to the README 🤔 Then again, this feels like *mostly* just rspec API, but we do seem to be adding a bit of behavior on top of this.

**Comment by AlexEvanczuk** [comment]
If we are requiring the user to add this new flag for rspec support, should we bump the major version? Or are we saying that not adding that flag, it continues to either work or not work as before, but adding that flag makes something new work? If so, it wouldn't be a breaking change.

File: README.md (hunk 1 above)
Comment (AlexEvanczuk) [reply README.md:96]: Also this is a big vague to me and something that might bother me a bit as a reader of this documentation. It seems like the "keep things simple" is more about allowing us to lean on the assumption that the user has done that to provide a simple internal implementation. If so, I don't care about that as a user of this and would just want to be told that I need to do this.

File: README.md [hunk 3]
```diff
@@ -92,6 +92,14 @@ metadata:
...
+Or, if you'd like, pass it as an argument to `rspec`:
+
+```
+$ rspec --require stimpack/rspec ...
```
Comment (AlexEvanczuk) [inline README.md:100]: Would we expect anyone to actually want to run rspec this way? I guess they could add that to a wrapper, just not sure why I'd want this.

File: lib/stimpack.rb [hunk 4]
```diff
@@ -44,6 +41,8 @@ def [](name)
     config/initializers
     config/routes
   )
+  # (0) / -> (1) /packs -> (2) /packs/foo/bar
+  @config.max_levels = 2
```
Comment (AlexEvanczuk) [inline lib/stimpack.rb:45]: Does this imply that folks can have N-level nested packs?

Comment (AlexEvanczuk) [reply lib/stimpack.rb:45]: And do we really want to support that? It won't work with the rest of the ruby at scale ecosystem.

File: lib/stimpack/integrations/rails.rb [hunk 5]
```diff
@@ -1,17 +1,53 @@
...
           Stimpack.config.paths.each do |path|
-            app.paths[path] << pack.path.join(path)
+            @app.paths[path] << pack.path.join(path)
           end
         end
       end
+
+      private
+
+      def create_namespace(name)
```
Comment (AlexEvanczuk) [inline lib/stimpack/integrations/rails.rb:34]: The implementation of these hasn't changed right? Just the location of the methods?

File: lib/stimpack/integrations/rspec.rb [hunk 6]
```diff
@@ -1,19 +1,46 @@
 module Stimpack
   module Integrations
     class RSpec
-      def self.install(app)
-        return unless defined?(::RSpec)
-        # Sometimes, the `rspec-rails` gem is installed in the development
```
Comment (AlexEvanczuk) [inline lib/stimpack/integrations/rspec.rb:6]: Nice to remove this comment and associated load order complexity.

File: lib/stimpack/integrations/rspec.rb [hunk 7]
```diff
@@ -1,19 +1,46 @@
...
-        # group. This means the ::RSpec module will be defined. However, this
-        # doesn't mean that we've loaded rspec-core, or even about to run tests.
-        # Let's make sure we are actually loading the test environment before
-        # installing our integration. An easy of doing this is seeing of the
-        # configuration method exists on RSpec.
-        return unless ::RSpec.respond_to?(:configuration)
+      def initialize
+        # This is the list of directories RSpec was told to run.
+        to_run = ::RSpec.configuration.instance_variable_get(:@files_or_directories_to_run)
```
Comment (AlexEvanczuk) [inline lib/stimpack/integrations/rspec.rb:6]: This kind of looks like private API of rspec -- are we comfortable relying on it?

File: lib/stimpack/integrations/rspec.rb [hunk 8]
```diff
@@ -1,19 +1,46 @@
...
+          #
+          # packs/my_pack => packs/my_pack/spec
+          # 
+          # If it doesn't match a pack path, we leave it alone.
+
+          to_run.map! do |path|
+            if pack = Packs.all_by_path[path]
+              [
+                pack,
+                *pack.all_children
```
Comment (AlexEvanczuk) [inline lib/stimpack/integrations/rspec.rb:32]: This is great -- I do think documenting this in the README would be helpful!

File: lib/stimpack/pack.rb [hunk 9]
```diff
@@ -1,53 +1,55 @@
...
     attr_reader :path
-    attr_reader :engine
+    attr_reader :name
+    attr_accessor :engine
 
-    def self.create(path)
-      new(path) if path.join(PACKAGE_FILE).exist?
-    end
+    def initialize(path, level:, packs_path: nil)
+      if Stimpack.config.max_levels && level > Stimpack.config.max_levels
```
Comment (AlexEvanczuk) [inline lib/stimpack/pack.rb:17]: Could/should we just hardcode and introduce this configurability when/if we want/need it?

**Review by AlexEvanczuk — APPROVED** [review]

File: lib/stimpack/integrations/rspec.rb (hunk 7 above)
Comment (ngan) [reply lib/stimpack/integrations/rspec.rb:6]: It is, but it hasn't changed for quite some time. I think I can find a better way or work with RSpec folks to get this right. In the meantime, we'll rely on this and we can ensure we can continue to support this.

File: lib/stimpack/integrations/rspec.rb (hunk 8 above)
Comment (ngan) [reply lib/stimpack/integrations/rspec.rb:32]: I'm not entirely sure which methods we want to make public/private yet.

File: lib/stimpack.rb (hunk 4 above)
Comment (ngan) [reply lib/stimpack.rb:45]: I'll remove this config.

File: lib/stimpack/integrations/rspec.rb (hunk 7 above)
Comment (AlexEvanczuk) [reply lib/stimpack/integrations/rspec.rb:6]: Sounds good. If they stop supporting it hopefully our tests break pretty dramatically 🤞

File: lib/stimpack/integrations/rspec.rb (hunk 8 above)
Comment (AlexEvanczuk) [reply lib/stimpack/integrations/rspec.rb:32]: This seems key though in terms of the behavior of rspec based on what args are passed in.

Comment (AlexEvanczuk) [reply lib/stimpack/integrations/rspec.rb:32]: I'm specifically referring to:

bundle exec rspec #=> Runs _all_ specs in your entire application (packs and all)
bundle exec rspec spec/some/specific_spec.rb #=> Runs just that one test
bundle exec rspec packs/foobar #=> Runs all specs under the foobar pack _and_ all the specs of its nested packs
bundle exec rspec packs/foobar pack/binbaz #=> same as above but also adds the binbaz pack
bundle exec rspec packs/foobar/spec #=> Runs all spec files inside the "packs/foobar/spec" directory
bundle exec rspec packs/foobar/nested_pack #=> Runs all specs under the "packs/foobar/nested_pack" pack

Comment (ngan) [reply lib/stimpack/integrations/rspec.rb]: Done!

File: README.md (hunk 3 above)
Comment (ngan) [reply README.md:100]: Not sure -- maybe CI?

File: spec/stimpack_spec.rb (hunk 2 above)
Comment (ngan) [reply spec/stimpack_spec.rb:4]: I don't think there's a way to "unload" a rails environment. I think we should just have the entire test suite run for a single Rails (version) environment and invoke the suite multiple times.

The `Dir.chdir` is necessary because we now don't have `Rails.root` and rely on Rails directory traversal. It works for now since we don't have anything relying on the current working dir. I'll try and think of a better way to find project root...but it works for now.

File: lib/stimpack/pack.rb (hunk 9 above)
Comment (ngan) [reply lib/stimpack/pack.rb]: This is no longer relevant

File: README.md (hunk 1 above)
Comment (ngan) [reply README.md]: Updated
